servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java

```diff
@@ -420,13 +422,29 @@ public void onComplete() {
                         }));
 
                 if (drainRequestPayloadBody) {
-                    responseWrite = responseWrite.afterFinally(() -> {
-                        if (!payloadSubscribed.get()) {
-                            // At this point, discarding the request payload body is an operation which should not
-                            // impact the state of request/response processing, and we need to do it in order to both
-                            // reuse the connection and finish any operators that may depend on both the request and
-                            // response bodies to have been fully processed.
-                            request.messageBody().ignoreElements().shareContextOnSubscribe().subscribe();
+                    responseWrite = responseWrite.afterFinally(new TerminalSignalConsumer() {
+                        @Override
+                        public void onComplete() {
+                            if (payloadSubscribed.compareAndSet(false, true)) {
+                                // At this point, discarding the request payload body is an operation which should not
+                                // impact the state of request/response processing, and we need to do it in order to
+                                // both reuse the connection and finish any operators that may depend on both the
+                                // request and response bodies to have been fully processed.
+                                request.messageBody().ignoreElements().shareContextOnSubscribe().subscribe();
+                            }
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            onComplete();
```

Comment:
This should be the same as `cancel()` to behave the same way as `Single.concatPropagateCancel(...)`. The typical flow for any other operator in case of an error is to propagate cancel